### PR TITLE
Adding transformers

### PR DESCRIPTION
**Pull request by zokost**
PublicTransformer - implemented class for transformers library

in detection_base_transformers.yaml you can see how the config should be modified

### CODE REVIEW

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    """Transformers interface."""
+    def __init__(
+            self,
+            emb_dim,
+            model):
+        super().__init__()
+        self.model = model
+        self.model.config
+        self._hidden_size = self.model.config.n_embd
+        self.delta_time = False
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:17]: Лучше оформить как свойство, так сразу видно интерфейс:
```python
@property
def delta_time(self):
   return False
```

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        p = next(iter(self.parameters()))
+        return torch.zeros(1, self._hidden_size, dtype=p.dtype, device=p.device)  # (1, D).
+
+    def forward(self, x: PaddedBatch, time_deltas: PaddedBatch,
+                states: Optional[Tensor]=None, return_states=False) -> Tuple[PaddedBatch, Tensor]:
+        """Apply Transformer.
+
+        Args:
+            x: Batch with shape (B, L, D).
+            time_deltas (unused): Relative inputs timestamps.
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:35]: Переписать на timestamps.

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            states: Initial states with shape (N, B, D), where N is the number of layers.
+            return_full_states: Whether to return full states with shape (B, L, D)
+                or only output states with shape (B, D).
+
+        Returns:
+            Outputs with shape (B, L, D) and states with shape (N, B, D) or (N, B, L, D), where
+            N is the number of layers.
+        """
+        x_emb = self.input_proj(x.payload)
+        outputs = self.model(inputs_embeds = x_emb)  # (B, L, D).
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:45]: (!) Нужно attention mask правильный сделать на основе x.seq_len_mask.

Comment (ivan-chai) [reply hotpp/nn/encoder/transformer/transformers_example.py:45]: Прошу проверить batch_first семантику.

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        elif return_states == "full":
+            output_states = all_hidden_states  # (N, B, L, D)
+        elif not return_states:
+            output_states = None
+        else:
+            raise ValueError(f"Unknown return_states flag: {return_states}")
+        
+        return PaddedBatch(outputs.last_hidden_state, x.seq_lens), output_states
+    
+    def interpolate(self, states: Tensor, time_deltas: PaddedBatch) -> PaddedBatch:
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:60]: Добавить raise NotImplementedError. Этот тип моделей нельзя использовать как continuous-time.

**Review by ivan-chai — CHANGES_REQUESTED** [review]

File: hotpp/nn/encoder/transformer/__init__.py [hunk 5]
```diff
@@ -1,2 +1,3 @@
 from .attnhp import AttNHPTransformer
 from .state import TransformerState
+from .transformers_example import HuggingFaceTransformer
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/__init__.py:3]: Название файла странное. Лучше huggingface или hf_model или hf_transformer. Там же не только трансформеры? Та же мамба ложится в интерфейс.

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        if return_states:
+            raise ValueError("Transformers encoder doesn't support states return")
+        
+        x_emb = self.input_proj(x.payload)
+        outputs = self.model(inputs_embeds = x_emb, attention_mask = x.seq_len_mask)  # (B, L, D).
+
+        return PaddedBatch(outputs.last_hidden_state, x.seq_lens), None
+
+    def interpolate(self, states: Tensor, time_deltas: PaddedBatch) -> PaddedBatch:
+        raise NotImplementedError('Transformers do not support continuous-time interpolation in this setup')
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:49]: В репе используются двойные кавычки)

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+class HuggingFaceTransformer(torch.nn.Module):
+    """Transformers interface."""
+    def __init__(
+            self,
+            input_size,
+            model):
+        super().__init__()
+        self.model = model
+        self._hidden_size = self.model.config.n_embd
+        self.input_proj = torch.nn.Linear(input_size, self.model.config.n_embd) #Layer to get right size
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:16]: Лучше не сокращать. input_projection вполне норм.

File: hotpp/nn/encoder/transformer/transformers_example.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            return_states: Whether to return final states with shape (B, D), full states with shape (B, T, D)
+                or no states (either False, "last" or "full"). Must be False. Exists only for interface compatibility.
+
+        Returns:
+            Outputs with shape (B, L, D) and None (states are not supported).
+        """
+        if return_states:
+            raise ValueError("Transformers encoder doesn't support states return")
+        
+        x_emb = self.input_proj(x.payload)
```
Comment (ivan-chai) [inline hotpp/nn/encoder/transformer/transformers_example.py:43]: Лучше не скоращать. embeddings, как вариант.

**Review by ivan-chai — CHANGES_REQUESTED** [review]
